### PR TITLE
Don't create empty parts on INSERT

### LINKED ISSUE #20304: optimize_on_insert creates empty parts
It would be great if `optimize_on_insert` was smart enough to discard empty blocks
```
create table empty (num Int8) engine=Memory;
CREATE MATERIALIZED VIEW empty_mv
(
    `num` Int8
)
ENGINE = SummingMergeTree
ORDER BY tuple() AS
SELECT num
FROM empty;

-- First insert

set optimize_on_insert=1;
insert into empty values (0);

-- 20 minutes break

set optimize_on_insert=0;
insert into empty values (0);

-- Created parts

SELECT
    event_type,
    event_time,
    table,
    part_name,
    rows
FROM system.part_log
PREWHERE (event_date = today()) AND (table LIKE '%empty%')
ORDER BY event_time ASC
FORMAT PrettyCompactMonoBlock

┌─event_type─┬──────────event_time─┬─table───────────┬─part_name─┬─rows─┐
│ NewPart    │ 2021-02-10 21:27:40 │ .inner.empty_mv │ all_1_1_0 │    0 │
│ RemovePart │ 2021-02-10 21:27:42 │ .inner.empty_mv │ all_1_1_0 │    0 │
│ NewPart    │ 2021-02-10 21:46:42 │ .inner.empty_mv │ all_2_2_0 │    1 │
└────────────┴─────────────────────┴─────────────────┴───────────┴──────┘
```

### PR DESCRIPTION
**Pull request by Avogar**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Do not create empty parts on INSERT when `optimize_on_insert` setting enabled. Fixes #20304.

Detailed description / Documentation draft:
If `optimize_on_insert` setting is true, the inserted block may become empty after merge. So, there is no need to create empty part.

### CODE REVIEW

**Review by alexey-milovidov — APPROVED** [review]
BTW, what about `tmp_` directory - will it be deleted immediately?

**Comment by alexey-milovidov** [comment]
About empty part after merge - should consult with @alesapin or @CurtizJ.

**Comment by alesapin** [comment]
> BTW, what about `tmp_` directory - will it be deleted immediately?

Let's just add an integration test and check.

File: tests/queries/0_stateless/01560_optimize_on_insert.sql [hunk 1]
```diff
@@ -33,3 +33,10 @@ INSERT INTO aggregating_merge_tree VALUES (1, 1, '2020-01-01'), (2, 1, '2020-01-
 SELECT * FROM aggregating_merge_tree ORDER BY key;
 DROP TABLE aggregating_merge_tree;
 
+SELECT 'Check creating empty parts';
+DROP TABLE IF EXISTS empty;
+CREATE TABLE empty (key UInt32, val UInt32, date Datetime) ENGINE=SummingMergeTree(val) PARTITION BY date ORDER BY key;
```
Comment (alesapin) [inline tests/queries/0_stateless/01560_optimize_on_insert.sql:38]: Need a test with the replicated table.

**Review by alesapin — APPROVED** [review]

**Comment by filimonov** [comment]
https://github.com/ClickHouse/ClickHouse/issues/20304

**Comment by alesapin** [comment]
`test_query_deduplication` -- I don't see how it's related to changes. Looks like a flaky test because failed only with address sanitizer.

**Comment by alesapin** [comment]
Hung check -- not related to changes.

**Comment by alesapin** [comment]
UB reported https://github.com/ClickHouse/ClickHouse/issues/21769

**Pull request merged**